### PR TITLE
Get value_shape from FunctionSpace

### PR DESCRIPTION
**Pull request by pbrubeck**
# Description

1. Replace `V.ufl_element().value_shape` by `V.value_shape`,
2. `firedrake.FunctionSpace.value_size` has been redefined to agree with `ufl.FunctionSpace.value_size = numpy.prod(value_shape, int)`, so the old behaviour gets renamed as `firedrake.FunctionSpace.block_size`

Depends on:
https://github.com/firedrakeproject/ufl/pull/52
https://github.com/FInAT/FInAT/pull/118
https://github.com/firedrakeproject/tsfc/pull/307

### CODE REVIEW

File: firedrake/interpolation.py [hunk 1]
```diff
@@ -1024,13 +1023,13 @@ def _interpolator(V, tensor, expr, subset, arguments, access, bcs=None):
...
+    if len(expr.ufl_shape) != len(V.value_shape):
         raise RuntimeError('Rank mismatch: Expression rank %d, FunctionSpace rank %d'
-                           % (len(expr.ufl_shape), len(V.ufl_element().value_shape)))
+                           % (len(expr.ufl_shape), len(V.value_shape)))
 
-    if expr.ufl_shape != V.ufl_element().value_shape:
+    if expr.ufl_shape != V.value_shape:
         raise RuntimeError('Shape mismatch: Expression shape %r, FunctionSpace shape %r'
-                           % (expr.ufl_shape, V.ufl_element().value_shape))
+                           % (expr.ufl_shape, V.ufl_element().value_shape(V.mesh())))
```
Comment (pbrubeck) [inline firedrake/interpolation.py:1032]:
```suggestion
                           % (expr.ufl_shape, V.value_shape))
```

File: firedrake/preconditioners/pmg.py [hunk 2]
```diff
@@ -1377,8 +1375,8 @@ def make_blas_kernels(self, Vf, Vc):
         # We could benefit from loop tiling for the transpose, but that makes the code
         # more complicated.
 
-        fshape = (Vf.value_size, Vf.finat_element.space_dimension())
-        cshape = (Vc.value_size, Vc.finat_element.space_dimension())
+        fshape = (numpy.prod(Vf.shape), Vf.finat_element.space_dimension())
+        cshape = (numpy.prod(Vc.shape), Vc.finat_element.space_dimension())
```
Comment (pbrubeck) [inline firedrake/preconditioners/pmg.py:1379]:
```suggestion
        fshape = (Vf.block_size, Vf.finat_element.space_dimension())
        cshape = (Vc.block_size, Vc.finat_element.space_dimension())
```

File: firedrake/constant.py [hunk 3]
```diff
@@ -79,7 +79,7 @@ def __new__(cls, value, domain=None, name=None, count=None):
 
             if not isinstance(domain, ufl.AbstractDomain):
                 cell = ufl.as_cell(domain)
-                coordinate_element = finat.ufl.VectorElement("Lagrange", cell, 1, gdim=cell.geometric_dimension)
```
Comment (pbrubeck) [inline firedrake/constant.py:82]: This doesn't look right

File: firedrake/pointquery_utils.py [hunk 4]
```diff
@@ -192,37 +193,46 @@ def predicate(index):
...
 @PETSc.Log.EventDecorator()
-def compile_coordinate_element(ufl_coordinate_element, contains_eps, parameters=None):
+def compile_coordinate_element(mesh, contains_eps, parameters=None):
     """Generates C code for changing to reference coordinates.
 
-    :arg ufl_coordinate_element: UFL element of the coordinates
-    :returns: C code as string
+    Parameters
+    ----------
+    mesh : MeshTopology
```
Comment (connorjward) [inline firedrake/pointquery_utils.py:201]:
```suggestion
    mesh :
```

File: firedrake/pointquery_utils.py [hunk 5]
```diff
@@ -192,37 +193,46 @@ def predicate(index):
 
 
 @PETSc.Log.EventDecorator()
-def compile_coordinate_element(ufl_coordinate_element, contains_eps, parameters=None):
+def compile_coordinate_element(mesh, contains_eps, parameters=None):
```
Comment (connorjward) [inline firedrake/pointquery_utils.py:196]:
```suggestion
def compile_coordinate_element(mesh: MeshGeometry, contains_eps: float, parameters: dict | None = None):
```
Note `MeshGeometry` not `MeshTopology` (I might be wrong though).

File: firedrake/pointquery_utils.py [hunk 6]
```diff
@@ -192,37 +193,46 @@ def predicate(index):
...
 
-    :arg ufl_coordinate_element: UFL element of the coordinates
-    :returns: C code as string
+    Parameters
+    ----------
+    mesh : MeshTopology
+        The mesh.
+
+    contains_eps : float
+        The tolerance used to verify that a point is contained by a cell.
```
Comment (connorjward) [inline firedrake/pointquery_utils.py:205]:
```suggestion
    contains_eps :
        The tolerance used to verify that a point is contained by a cell. 
    parameters :
        Form compiler parameters, defaults to whatever TSFC defaults to.
```

File: tests/regression/test_2dcohomology.py [hunk 7]
```diff
@@ -94,8 +94,8 @@ def test_betti1(space, mesh):
     L = assemble((inner(sigma, tau) - inner(u, rot(tau)) + inner(rot(sigma), v)
                   + inner(div(u), div(v))) * dx)
 
-    bc0 = DirichletBC(W.sub(0), Constant(0.0), 9)
-    bc1 = DirichletBC(W.sub(1), Constant((0.0, 0.0)), 9)
```
Comment (connorjward) [inline tests/regression/test_2dcohomology.py:98]: Why this change?

File: tests/regression/test_ensembleparallelism.py [hunk 8]
```diff
@@ -347,7 +347,7 @@ def test_send_and_recv(ensemble, mesh, W, blocking):
     parallel_assert(
         lambda: error < 1e-12,
         subset=root_ranks,
-        msg=f"{error = :.5f}"
+        msg=f"{error=:.5f}"
```
Comment (connorjward) [inline tests/regression/test_ensembleparallelism.py:350]: This may lead to conflicts as I fixed this exact issue yesterday.

**Review by connorjward — CHANGES_REQUESTED** [review]
Generally this looks great. @ksagiyam can you also give this a scan to make sure it's the right thing.

File: tests/regression/test_2dcohomology.py (hunk 7 above)
Comment (pbrubeck) [reply tests/regression/test_2dcohomology.py:98]: Nothing wrong here, just had to deal with similar bcs on tests, where I replaced `zero(V.ufl_element().value_shape)` with `0`. Prescribing a `Constant` will require to generate code, so this gives us a tiny speedup

File: firedrake/assemble.py [hunk 9]
```diff
@@ -1628,7 +1628,7 @@ def _integral_type(self):
 
     @cached_property
     def _mesh(self):
-        return self._form.ufl_domains()[self._kinfo.domain_number]
```
Comment (pbrubeck) [inline firedrake/assemble.py:1631]: @ksagiyam `ufl_domains()` returns a `set`, so I casted it into a tuple, how is this going to affect `MixedMesh`?

Comment (connorjward) [reply firedrake/assemble.py:1631]: That seems wrong. Can the set be sorted in some way?

Comment (ksagiyam) [reply firedrake/assemble.py:1631]: I think this has been fixed recently:
https://github.com/FEniCS/ufl/blob/d8350119ac162caf949f8aad6cee76223a5c637f/ufl/domain.py#L201

I will need to rebase my branch, but you do not need to worry about that.

Comment (pbrubeck) [reply firedrake/assemble.py:1631]: I encountered the issue with SLATE tensors (fix in this PR), and maybe FormSum ([fix in UFL](https://github.com/FEniCS/ufl/pull/325)). join_domains returns a set, so we need to call sort_domains afterwards,

**Review by ksagiyam — CHANGES_REQUESTED** [review]
Thanks!

File: firedrake/interpolation.py [hunk 10]
```diff
@@ -614,7 +614,7 @@ def __init__(
         # I first point evaluate my expression at these locations, giving a
         # P0DG function on the VOM. As described in the manual, this is an
         # interpolation operation.
-        shape = V_dest.ufl_element().value_shape
+        shape = V_dest.ufl_function_space().value_shape
```
[the remark this answers is not in the document]
Comment (pbrubeck) [reply firedrake/interpolation.py:617]: `V_dest` could be either a `Function` or a `FunctionSpace`

[the remark this answers is not in the document]
Comment (ksagiyam) [reply firedrake/interpolation.py:617]: OK.

**Review by ksagiyam — APPROVED** [review]

**Comment by ksagiyam** [comment]
Tests passed: https://github.com/firedrakeproject/firedrake/actions/runs/11858718532/job/33049993186?pr=3862
